Use fewer mocks in the association tests

markstory opened the pull request:
Extensive mocking seemed useful in the association tests initially. However, over time I feel it has made it harder for contributors to add tests, and caused many tests that should go into the association test cases, to end up in `QueryRegressionTest` or `TableTest`. If people are comfortable with these changes I'd like to:
- Cleanup the BelongsToManyTest.
- Move tests out of QueryRegression and TableTest into more appropriate locations.

markstory committed:
Add docs for HasMany saveStrategy.

Refs cakephp/docs#3553

markstory committed:
Use fewer mocks in BelongsTo tests.

We don't need this many mocks, and using concrete objects gives more
realistic test scenarios.

markstory committed:
Reduce number of mocks used in HasMany tests.

markstory committed:
Reduce number of mocks used in HasMany tests.

jeremyharris commented:
I like this. Less mocks where appropriate is a good thing :)

jeremyharris commented on tests/TestCase/ORM/Association/BelongsToTest.php:
```diff
@@ -181,26 +177,80 @@ public function testAttachTo()
+    public function testAttachToMultiPrimaryKey()
+    {
+        $this->company->primaryKey(['id', 'tenant_id']);
+        // $query = $this->getMock('\Cake\ORM\Query', ['join', 'select'], [null, null]);
```
Leftover mock statement

markstory replied:
Thanks :+1:

jeremyharris commented on tests/TestCase/ORM/Association/HasManyTest.php:
```diff
@@ -162,31 +174,32 @@ public function testEagerLoader()
+        $query = $this->article->query();
+        $this->article->method('find')
+            ->with('all')
             ->will($this->returnValue($query));
```
Are these leftover mock methods or is the diff confusing me?

ADmad commented:
Tests failing

lorenzo commented:
I like this :+1:

markstory commented:
@ADmad Identifier quoting tests got me again :cry:

markstory committed:
Fix tests that were failing due to identifier quoting.

Identifier quoting makes comparing expressions harder. Add some helper
methods to do assertions. Overtime these can be extracted as more
classes need them.

markstory committed:
Fix association casing.

dereuromark commented:
:clap:
